doc/design/eventdriven-callbacks.md

```diff
@@ -0,0 +1,271 @@
+      void onSampleReceived2(iox::popo::UntypedSubscriber & subscriber) {}
+
+      void onSampleReceived(iox::popo::UntypedSubscriber & subscriber) {
+        myCallbackReactAL.assignEvent(subscriber, iox::popo::SubscriberEvent::HAS_SAMPLE_RECEIVED, onSampleReceived2);
+      }
+
+      myCallbackReactAL.assignEvent(subscriber, iox::popo::SubscriberEvent::HAS_SAMPLE_RECEIVED, onSampleReceived);
+      ```
+ - One can assign at most one callback to a specific event of an object. The event is usually defined 
+    with an enum by the developer. One example is `SubscriberEvent::HAS_SAMPLE_RECEIVED`.
+    - Attaching a callback to an event where a callback has been already assigned overrides
+      the existing callback with the new one.
+ - concurrently: detach a callback at any time from anywhere. This means in particular
+    - Detaching a callback from within a callback.
+    - [robust] When the detach call returns we guarantee that the callback is never called
+      again. E.g. blocks till the callback is removed, if the callback is concurrently 
+      running it will block until the callback is finished and removed.
+      Exception: If a callback detaches itself it blocks until the callback is removed 
+                 and not until the callback is finished. After the successful removal the 
+                 callback will continue and will not be called again.
+    - Calling detach means that after that call the callback is no longer assigned
+      even when it was not assigned in the first place. Therefore it will always succeed.
+
+ - When the ReactAL goes out of scope it should detach itself from every class 
+     to which it was assigned via a callback provided by the attached class (like in the WaitSet).
+
+ - When the class which is assigned to the ReactAL goes out of scope it should 
+    detach itself from the ReactAL via a callback provided by the ReactAL (like in the WaitSet).
+
+### Usage Code Example
+```cpp
+ReactAL myCallbackReactAL;
+iox::popo::UntypedSubscriber mySubscriber;
+
+void onSampleReceived(iox::popo::UntypedSubscriber & subscriber) {
+  subscriber.take().and_then([&](auto & sample){
+    std::cout << "received " << std::hex << sample->payload() << std::endl;
+  });
+}
+
+void onWaitForSubscription(iox::popo::UntypedSubscriber & subscriber) {
+  std::cout << "subscribed!\n";
+  myCallbackReactAL.assignEvent(subscriber, iox::popo::SubscriberEvent::HAS_SAMPLE_RECEIVED, onSampleReceived);
+  myCallbackReactAL.assignEvent(subscriber, iox::popo::SubscriberEvent::UNSUBSCRIBED, onWaitForUnsubscription);
+  myCallbackReactAL.cancelEvent(subscriber, iox::popo::SubscriberEvent::SUBSCRIBED);
+}
+
+void onWaitForUnsubscribe(iox::popo::UntypedSubscriber & subscriber) {
+  std::cout << "unsubscribed from publisher\n";
+  myCallbackReactAL.assignEvent(subscriber, iox::popo::SubscriberEvent::SUBSCRIBED, onWaitForSubscription);
+  myCallbackReactAL.cancelEvent(subscriber, iox::popo::SubscriberEvent::HAS_SAMPLE_RECEIVED);
```

Comment:
Not sure about `cancelEvent`. I would not assume that it's also detached. Why not just keeping `detach`? `retract` or `retrieve` might also work better that `cancel`.

Reply:
I thought assign & cancel and attach & detach. Detach did sound like the opposite of assign to me. But when it is clearer for the rest of the world I will keep detach.

Reply:
maybe it's just me, but although it feels natural to say `cancel an assignment` it feels unnatural to me to use `cancel an event` with the meaning of not being assigned anymore.